File: common/src/main/java/dev/wren/truss/platform/Services.java
package dev.wren.truss.platform;

import dev.wren.truss.TrussCommon;
import dev.wren.truss.platform.services.PlatformHelper;

import java.util.ServiceLoader;

public class Services {

    public static final PlatformHelper PLATFORM = load(PlatformHelper.class);

    public static <T> T load(Class<T> clazz) {

        final T loadedService = ServiceLoader.load(clazz)
                .findFirst()
                .orElseThrow(() -> new NullPointerException("Failed to load service for " + clazz.getName()));
        TrussCommon.LOGGER.debug("Loaded {} for service {}", loadedService, clazz);
        return loadedService;
    }
}

File: common/src/main/kotlin/dev/wren/truss/platform/Services.kt
package dev.wren.truss.platform

import dev.wren.truss.Truss
import dev.wren.truss.platform.services.PlatformHelper
import java.util.ServiceLoader

val PLATFORM: PlatformHelper = load(PlatformHelper::class.java)

fun <T> load(clazz: Class<T>): T {
    val loadedService = ServiceLoader.load<T>(clazz, clazz.classLoader)
        .findFirst()
        .orElseThrow { NullPointerException("Failed to load service for " + clazz.getName()) }
    Truss.LOGGER.debug("Loaded {} for service {}", loadedService, clazz)
    return loadedService
}